refactor: add methods in errors package to quit with exit codes

darshanime opened the pull request:
This is a part of Streamline exit codes of CLI (https://github.com/argoproj/argo-cd/issues/3592) issue on argo-cd.  
Continued from where @jannfis [left off](https://github.com/jannfis/argo-cd/commit/4b5574ed8509098476a4bf672a5c7e8020e82b6a)

Signed-off-by: darshanime <deathbullet@gmail.com>

ash2k commented on agent/main.go:
```diff
@@ -119,10 +119,10 @@ func newCmd() *cobra.Command {
 			s := settings{args[0], paths}
 			config, err := clientConfig.ClientConfig()
-			errors.CheckError(err)
+			errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
```
I think this can be simplified if line 115 uses `RunE` field instead of `Run`. Then this function can just return the error directly and the checking logic can be centralized in `main()`.

darshanime replied:
`CheckErrorWithCode` just exits with the appropriate exit code (instead of the default exit code) -- there is no real error checking logic, we simply exit.

jannfis replied:
Moving the logic of error handling to `main()` would assume that the `error` would actually hold non-generic information about the error, but I think currently most `error` variables are just generics, with nothing more than an error message produced by `fmt.Errorf()` or similar methods. But I agree, in the long run, we should have specialised `error` variables that enables to differentiate the errors directly from the `error` variable.

ash2k replied:
I agree with the above. I just wanted to say that current approach might be less than ideal because:
- `defer`ed functions are not executed. Currently there is only one and it might not be a big deal. This might change in the future.
- Exiting on error might need something that is not accessible from a static method (e.g. a field or method on some object). E.g. the program may want to send an event to a metric/error aggregation system. Having multiple places that do this (vs one in the `main()`) might be more verbose.
- Having 1 entry point and 1 exit point is easier to follow in general, but this is subjective.

jannfis commented on agent/main.go:
```diff
@@ -162,7 +162,7 @@ func newCmd() *cobra.Command {
 			})
 			go func() {
-				errors.CheckError(http.ListenAndServe(fmt.Sprintf("0.0.0.0:%d", port), nil))
+				errors.CheckErrorWithCode(http.ListenAndServe(fmt.Sprintf("0.0.0.0:%d", port), nil), errors.ErrorConnectionFailure)
```
I think this is not a `ErrorConnectionFailure` class of error, since it creates a listener and does not try to establish a connection to the API.

jannfis commented on pkg/cache/cluster_test.go:
```diff
@@ -26,7 +26,7 @@ import (
 	obj := make(map[string]interface{})
 	err := yaml.Unmarshal([]byte(jsonStr), &obj)
-	errors.CheckError(err)
+	errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
```
I think the tests can just exit with any exit code and do not need adaption. They should just use `CheckError()` or even better, `panic()` (see below).

jannfis commented on pkg/diff/diff_test.go:
```diff
@@ -56,16 +56,16 @@ func toUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
-	errors.CheckError(err)
+	errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
 	return un
 }
 
 func unmarshalFile(path string) *unstructured.Unstructured {
 	data, err := ioutil.ReadFile(path)
-	errors.CheckError(err)
+	errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
 	var un unstructured.Unstructured
 	err = json.Unmarshal(data, &un.Object)
-	errors.CheckError(err)
+	errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
```
Same here as with above - tests should just use `CheckError()` (or even better, I think `panic()` would be the right thing to do in tests, but we should think about this -- CC @alexmt)

jannfis commented on pkg/utils/errors/errors.go:
```diff
@@ -1,18 +1,55 @@
 
-// CheckError is a convenience function to exit if an error is non-nil and exit if it was
-func CheckError(err error) {
+const (
+	// ErrorCommandSpecific is reserved for command specific indications
+	ErrorCommandSpecific = 1
+	// ErrorConnectionFailure is returned on connection failure to API endpoint
+	ErrorConnectionFailure = 11
+	// ErrorAPIResponse is returned on unexpected API response, i.e. authorization failure
+	ErrorAPIResponse = 12
+	// ErrorResourceDoesNotExist is returned when the requested resource does not exist
+	ErrorResourceDoesNotExist = 13
+	// ErrorGeneric is returned for generic error
+	ErrorGeneric = 20
+)
```
We should think about whether these codes are sufficient and if the classifications are correct. The code where I left off was clearly not finished yet and not thought through to the end, so I'm unsure whether we should adapt it in this state in a verbatim way

darshanime replied:
I have used these 5 error classes for argo-cd in https://github.com/argoproj/argo-cd/pull/3700 and it seems okay. I guess we'll truly know if they are sufficient in you can review that PR 😅

darshanime commented:
@jannfis thanks for your review. I have addressed the comments, ptal. 

Agree that the methods should return the correct error and `main()` have the centralized error handling logic. I can refactor that in a subsequent PR. In this one, I wanted to limit the scope to adding the helper methods to exit with input error codes. Let me know if it'll be better to refactor here itself.

ash2k commented on agent/main.go:
```diff
@@ -141,10 +141,10 @@ func newCmd() *cobra.Command {
 			)
 			gitOpsEngine := engine.NewEngine(config, clusterCache)
-			errors.CheckError(err)
+			errors.CheckErrorWithCode(err, errors.ErrorCommandSpecific)
```
Was looking at the code and noticed that this check is not needed. Trying to use the engine...
